src/irmin-pack/inode.ml

```diff
@@ -50,680 +50,664 @@ struct
+          match t.v with
+          | Values vs ->
+              let length =
+                match v' with
+                | None -> StepMap.cardinal vs + 1
+                | Some _ -> StepMap.cardinal vs
+              in
+              let t =
+                if length <= Conf.entries then values (StepMap.add s v vs)
+                else
+                  let vs = StepMap.bindings (StepMap.add s v vs) in
+                  let empty =
+                    inodes
+                      {
+                        length = 0;
+                        seed;
+                        entries = Array.make Conf.entries Empty;
+                      }
+                  in
+                  let aux t (s, v) =
+                    (add [@tailcall]) ~seed ~find ~copy:false t s v (fun x -> x)
+                  in
+                  List.fold_left aux empty vs
+              in
+              k t
           | Inodes t -> (
+              let length =
+                match v' with None -> t.length + 1 | Some _ -> t.length
+              in
+              let entries = if copy then Array.copy t.entries else t.entries in
               let i = index ~seed s in
-              let x = t.entries.(i) in
-              match x with
-              | Empty -> None
-              | Inode i -> aux ~seed:(seed + 1) (get_tree ~find i).v)
-        in
-        aux ~seed t.v
-
-      let find ~find t s = find_value ~seed:0 ~find t s
-
-      let rec add ~seed ~find ~copy t s v k =
-        match find_value ~seed ~find t s with
-        | Some v' when equal_value v v' -> k t
-        | v' -> (
-            match t.v with
-            | Values vs ->
-                let length =
-                  match v' with
-                  | None -> StepMap.cardinal vs + 1
-                  | Some _ -> StepMap.cardinal vs
-                in
-                let t =
-                  if length <= Conf.entries then values (StepMap.add s v vs)
-                  else
-                    let vs = StepMap.bindings (StepMap.add s v vs) in
-                    let empty =
-                      inodes
-                        {
-                          length = 0;
-                          seed;
-                          entries = Array.make Conf.entries Empty;
-                        }
-                    in
-                    let aux t (s, v) =
-                      (add [@tailcall]) ~seed ~find ~copy:false t s v (fun x ->
-                          x)
-                    in
-                    List.fold_left aux empty vs
-                in
+              match entries.(i) with
+              | Empty ->
+                  let tree = values (StepMap.singleton s v) in
+                  entries.(i) <- inode ~tree tree.hash;
+                  let t = inodes { seed; length; entries } in
+                  k t
+              | Inode n ->
+                  let t = get_tree ~find n in
+                  (add [@tailrec]) ~seed:(seed + 1) ~find ~copy t s v
```

Comment:
IIRC the tailcall annotations don't always work properly on `flambda` switches so we ended up removing them.

Reply:
This one should be fine since it can't be inlined: a greater issue is that it's going to be ignored because the annotation is called `tailcall` and not `tailrec` :stuck_out_tongue:

The issue being referred to is here: https://github.com/ocaml/ocaml/issues/9510.

Reply:
Ooooooops 👍

Reply:
I just remove the diff then